presto-main/src/main/java/com/facebook/presto/type/TimeOperators.java

```diff
@@ -98,7 +107,25 @@ public static boolean between(@SqlType(StandardTypes.TIME) long value, @SqlType(
     @SqlType(StandardTypes.TIME_WITH_TIME_ZONE)
     public static long castToTimeWithTimeZone(ConnectorSession session, @SqlType(StandardTypes.TIME) long value)
     {
-        return packDateTimeWithZone(value, session.getTimeZoneKey());
+        if (session.isLegacyTimestamp()) {
+            return packDateTimeWithZone(value, session.getTimeZoneKey());
+        }
+        else {
+            // FIXME This is hack that we need to use as the timezone interpretation depends on date (not only on time)
+            long currentMillisOfDay =
+                    LocalTime.from(
+                            Instant.ofEpochMilli(REFERENCE_TIMESTAMP_UTC)
+                                    .atZone(ZoneId.of(TimeZoneKey.UTC_KEY.getId())))
+                            .toNanoOfDay() / 1_000_000L; // nanos to millis
+
+            long timeMillisUtcInCurrentDay = REFERENCE_TIMESTAMP_UTC - currentMillisOfDay + value;
+
+            ISOChronology localChronology = getChronology(session.getTimeZoneKey());
+
+            // This cast does treat TIME as wall time in session TZ. This means that in order to get
+            // its UTC representation we need to shift the value by the offset of TZ.
+            return packDateTimeWithZone(value - localChronology.getZone().getOffset(timeMillisUtcInCurrentDay), session.getTimeZoneKey());
```

Comment:
I fail to understand why `return packDateTimeWithZone(value, session.getTimeZoneKey());` wasn't good. I think you must fix me before I go any further